// 0g/dist/ComponentManager.js
import { ComponentPool } from './ComponentPool.js';
import { COMPONENT_CHANGE_HANDLE, } from './Component2.js';
/**
 * Manages pools of Components based on their Type, and
 * the presence of Components assigned to Entities.
 */
export class ComponentManager {
    constructor(componentHandles, game) {
        this.componentHandles = componentHandles;
        this.game = game;
        this.pools = new Array();
        this.changed = new Array();
        this.unsubscribes = new Array();
        this.acquire = (typeId, initialValues) => {
            if (!this.pools[typeId]) {
                throw new Error(`ComponentType with ID ${typeId} does not exist`);
            }
            const component = this.pools[typeId].acquire(initialValues, this.game.idManager.get());
            component.$[COMPONENT_CHANGE_HANDLE] = this.onComponentChanged;
            return component;
        };
        this.release = (instance) => {
            delete instance.$[COMPONENT_CHANGE_HANDLE];
            return this.pools[instance.$.type.id].release(instance);
        };
        this.wasChangedLastFrame = (componentInstanceId) => {
            return !!this.changed[componentInstanceId];
        };
        this.onComponentChanged = (component) => {
            this.game.enqueueOperation({
                op: 'markChanged',
                componentId: component.$.id,
            });
        };
        this.markChanged = (component) => {
            this.changed[component.id] = true;
        };
        this.resetChanged = () => {
            this.changed.length = 0;
        };
        this.getTypeName = (typeId) => {
            return this.pools[typeId].ComponentType.name;
        };
        this.destroy = () => {
            this.unsubscribes.forEach((unsub) => unsub());
            this.pools.forEach((pool) => pool.destroy());
        };
        // initialize pools, one for each ComponentType by ID. ComponentType IDs are incrementing integers.
        Object.values(componentHandles).forEach((handle) => {
            // assign an ID
            handle.id = game.idManager.get();
            // create a pool
            this.pools[handle.id] = new ComponentPool(handle, this.game);
        });
        // TODO: right time to do this?
        this.unsubscribes.push(game.subscribe('preApplyOperations', this.resetChanged));
    }
}
