// 0g/dist/input/keyboard.js
export class Keyboard {
    constructor() {
        this.keysPressed = new Set();
        this.keysDown = new Set();
        this.keysUp = new Set();
        this.handleKeyDown = (ev) => {
            if (ev.target === document.body) {
                ev.preventDefault();
            }
            const key = ev.key;
            // avoid key-repeat triggering?
            if (!this.keysPressed.has(key)) {
                this.keysPressed.add(key);
                this.keysDown.add(key);
            }
        };
        this.handleKeyUp = (ev) => {
            const key = ev.key;
            this.keysPressed.delete(ev.key);
            this.keysUp.add(key);
        };
        this.getKeyPressed = (key) => {
            return this.keysPressed.has(key);
        };
        this.getKeyDown = (key) => {
            return this.keysDown.has(key);
        };
        this.getKeyUp = (key) => {
            return this.keysUp.has(key);
        };
        this.getAllPressedKeys = () => {
            return this.keysPressed;
        };
        this.frame = () => {
            this.keysDown.clear();
            this.keysUp.clear();
        };
        window.addEventListener('keydown', this.handleKeyDown);
        window.addEventListener('keyup', this.handleKeyUp);
    }
}
export const keyboard = new Keyboard();
